### ISSUE TITLE
The `set server state` unix socket command can't find all the servers

### ISSUE THREAD

**Issue by x-yuri** [haproxy/haproxy#2620]
### Detailed Description of the Problem

At least in [my minimal case][a] (which is actually steps to reproduce the issue) where I define 4 servers using `server-template`:

```
$ echo set server http/s1 state ready | ...
No such server.
$ echo set server http/s2 state ready | ...
```

The same happens with [`master`][b], but not with the `haproxy:2.9-alpine` image.

[a]: https://gist.github.com/x-yuri/75e1cc668c5b5951c7e3f1e4c01f1921
[b]: https://github.com/haproxy/haproxy/commit/ad946a704dc19b1a5aa51692ca7aafb5b015ba7c

### Expected Behavior

The command succeeding for every server.

### Steps to Reproduce the Behavior

See the [gist][a].

### Do you have any idea what may have caused this?

My research shows that [here][c] `strdup()` overwrites `s1` in the `http_proxy->conf.used_server_name` tree, and then [`cli_find_server()`][d] can't find it. I may be wrong, but that's what I see.

[c]: https://github.com/haproxy/haproxy/blob/v3.0.0/src/resolvers.c#L3495
[d]: https://github.com/haproxy/haproxy/blob/v3.0.0/src/server.c#L5067

### What is your configuration?

```haproxy
global
    maxconn  1024
    stats socket /var/lib/haproxy/haproxy.sock level admin

listen http
    mode http
    bind *:80
    server-template s 4 app:80 check init-addr last,none resolvers default
    timeout connect 5s
    timeout client 50s
    timeout server 50s
    option httpchk GET /health
```

### Output of `haproxy -vv`

```plain
HAProxy version 3.0.2-a45a8e6 2024/06/14 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2029.
Known bugs: http://www.haproxy.org/bugs/bugs-3.0.2.html
Running on: Linux 6.9.1-arch1-1 #1 SMP PREEMPT_DYNAMIC Fri, 17 May 2024 16:56:38 +0000 x86_64
Build options :
  TARGET  = linux-musl
  CC      = cc
  CFLAGS  = -O2 -g -fwrapv
  OPTIONS = USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_PROMEX=1 USE_PCRE2=1 USE_PCRE2_JIT=1
  DEBUG   = 

Feature list : -51DEGREES +ACCEPT4 -BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_CAP +LINUX_SPLICE +LINUX_TPROXY +LUA +MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_AWSLC -OPENSSL_WOLFSSL -OT -PCRE +PCRE2 +PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION -QUIC -QUIC_OPENSSL_COMPAT +RT +SHM_OPEN +SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 -SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL -ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=4).
Built with OpenSSL version : OpenSSL 3.3.1 4 Jun 2024
Running on OpenSSL version : OpenSSL 3.3.1 4 Jun 2024
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with Lua version : Lua 5.4.6
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with libslz for stateless compression.
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.43 2024-02-16
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 13.2.1 20240309

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[BWLIM] bwlim-in
	[BWLIM] bwlim-out
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[SPOE] spoe
	[TRACE] trace
```

**Comment by Darlelet** [comment]
Thanks for the report, I'm able to reproduce

First bad commit is:
b4db3be86e03ec31055303d46056223bb989472e

7223296092c72c2b10729a6da218897887858dba aimed at fixing the above issue, but it is not sufficient and leads to UAF when searching for the first server by name because the original ID gets replaced by `_srv_parse_set_id_from_prefix()` but we don't update `used_server_name` tree node consequently (which relies on srv->id pointer).

The fix below should help:
```diff
diff --git a/src/server.c b/src/server.c
index caf2f40bb..99a43eb5b 100644
--- a/src/server.c
+++ b/src/server.c
@@ -3177,7 +3177,10 @@ static int _srv_parse_tmpl_init(struct server *srv, struct proxy *px)
 		newsrv->conf.name.key = newsrv->id;
 		ebis_insert(&curproxy->conf.used_server_name, &newsrv->conf.name);
 	}
+	ebpt_delete(&srv->conf.name);
 	_srv_parse_set_id_from_prefix(srv, srv->tmpl_info.prefix, srv->tmpl_info.nb_low);
+	srv->conf.name.key = srv->id;
+	ebis_insert(&curproxy->conf.used_server_name, &srv->conf.name);
 
 	return i - srv->tmpl_info.nb_low;

```

**Comment by Darlelet** [comment]
Actually the `_srv_parse_set_id_from_prefix()` at the end of `_srv_parse_tmpl_init()` is not needed anymore since 7223296, it was probably overlooked. Thus the fix is even simpler:

```diff
diff --git a/src/server.c b/src/server.c
index caf2f40bb..5c2575e33 100644
--- a/src/server.c
+++ b/src/server.c
@@ -3177,7 +3177,6 @@ static int _srv_parse_tmpl_init(struct server *srv, struct proxy *px)
 		newsrv->conf.name.key = newsrv->id;
 		ebis_insert(&curproxy->conf.used_server_name, &newsrv->conf.name);
 	}
-	_srv_parse_set_id_from_prefix(srv, srv->tmpl_info.prefix, srv->tmpl_info.nb_low);
 
 	return i - srv->tmpl_info.nb_low;
 ```

**Comment by Darlelet** [comment]
Meanwhile (waiting for the fix in the next 3.0 release), a good workaround would be to target the server using numerical id (`#` followed by id instead of name), since this only happens with the first server:
```
set server http/#1 state ready
```

**Comment by x-yuri** [comment]
What did I mean by "[`strdup()`][b] overwrites `s1` in the `http_proxy->conf.used_server_name` tree"? In the [following gist][a] I added a couple of lines before and after the `strdup`. The output:

```
// displaying the tree
haproxy-1  | -- resolvers_new: http
haproxy-1  | s1
haproxy-1  | s2
haproxy-1  | s2
haproxy-1  | s3
haproxy-1  | s3
haproxy-1  | s4
haproxy-1  | s4

// displaying the value in the s1 node (key)
haproxy-1  | s1: s1

// the address of the s1 node's key:
haproxy-1  | 0x5db008ee9940
// the address returned by strdup():
haproxy-1  | 0x5db008ee9940

// displaying the tree
haproxy-1  | -- resolvers_new: http
haproxy-1  | default
haproxy-1  | s2
haproxy-1  | s2
haproxy-1  | s3
haproxy-1  | s3
haproxy-1  | s4
haproxy-1  | s4

// displaying the s1 node
haproxy-1  | s1: default
```

In other words the key in the `s1` node changes from `s1` to `default` (which is apparently the name of the resolvers section being created). And the address of the `s1` node's key matches the address returned by `strdup()`. It looks to me like `strdup()` considers the memory occupied by the `s1` node as free and overwrites it. Unless I made some mistake.

[a]: https://gist.github.com/x-yuri/8a0d063370761ee0e44e34289d12aaa9
[b]: https://github.com/haproxy/haproxy/blob/v3.0.0/src/resolvers.c#L3495

**Comment by Darlelet** [comment]
What you see is probably the result of UAF (use-after-free) since the memory location may be reused for something else once it is freed, but the above fix addresses the root cause and should fix your issue. I'll merge it soon and it should be available in the next 3.0 release
